bokeh/core/has_props.py

```diff
@@ -607,8 +607,16 @@ def apply_theme(self, property_values):
 
         # Emit any change notifications that result
         for k, v in old_values.items():
-            prop = self.lookup(k)
-            prop.trigger_if_changed(self, v)
+            descriptor = self.lookup(k)
+            if isinstance(v, PropertyValueContainer) and v._unmodified_default_value:
+                if k in self._property_values:
+                    self._property_values[k]._unregister_owner(self, descriptor)
+                    del self._property_values[k]
+                if k in property_values:
+                    self._property_values[k] = descriptor.property.prepare_value(self.__class__, k, property_values[k])
+                    self._property_values[k]._unmodified_default_value = True
+                    self._property_values[k]._register_owner(self, descriptor)
+            descriptor.trigger_if_changed(self, v)
```

Comment:
If I'm reading the code correctly, this is effectively a cut-and-paste of logic that's elsewhere, but the "elsewhere" is not transparently identical, it's spread around some different methods and moreover kind of tricky.  Chances of changing this and the elsewhere separately, or having them work differently even to start, seem very high.

If I understand right, the problem is a failure of cache invalidation. For containers, we want to make a new container per-instance and then cache it, so that mutating the container won't affect all instances.

The issue is that when we apply a new theme, values from the old theme are left over in `_property_values` and treated as if the application set them.

Would it work to add code right before this for loop (right after setting the new `__themed_values__`) which would "invalidate the cache" in _property_values, I believe something like this:
```
# Delete any cached values that came from the old theme
for k in old_dict.keys():
   v = old_values.get(k, None)
   # not sure the "k in self._property_values" check is needed
   if v is not None and isinstance(v, PropertyValueContainer) and \
        v._unmodified_default_value and k in self._property_values:
     del self._property_values[k]
```

Now as you go through and `trigger_if_changed`, the trigger_if_changed code should getattr using the _new_ theme, which should trigger caching the new value in `_property_values` and doing the registration and all that using the existing codepath?

Wouldn't be surprised if there's some way this fails but it seems like more of a "root cause" fix rather than patching it up later.